dice_ml/explainer_interfaces/dice_KD.py

```diff
@@ -0,0 +1,236 @@
+            self.total_CFs = total_CFs  # size of counterfactual set
+
+        # freeze those columns that need to be fixed
+        if features_to_vary != self.features_to_vary:
+            self.features_to_vary = features_to_vary
+            self.feat_to_vary_idxs = self.data_interface.get_indexes_of_features_to_vary(features_to_vary=features_to_vary)
+            self.freezer = tf.constant([1.0 if ix in self.feat_to_vary_idxs else 0.0 for ix in range(len(self.minx[0]))])
+
+
+        # CF initialization
+        if len(self.cfs) != self.total_CFs:
+            self.cfs = []
+            for ix in range(self.total_CFs):
+                one_init = [[]]
+                for jx in range(self.minx.shape[1]):
+                    one_init[0].append(np.random.uniform(self.minx[0][jx], self.maxx[0][jx]))
+                self.cfs.append(tf.Variable(one_init, dtype=tf.float32))
+
+    def find_counterfactuals(self, query_instance, desired_class, optimizer, learning_rate, min_iter, max_iter, project_iter, loss_diff_thres, loss_converge_maxiter, verbose, init_near_query_instance, tie_random, stopping_threshold, posthoc_sparsity_param, posthoc_sparsity_algorithm):
+        """Finds counterfactuals by gradient-descent."""
+
+        # Prepares user defined query_instance for DiCE.
+        query_instance_ = self.data_interface.prepare_query_instance(query_instance=query_instance, encode=True)
+        query_instance_ = np.array([query_instance_.iloc[0].values])
+
+        # find the predicted value of query_instance
+        test_pred = self.predict_fn(tf.constant(query_instance_, dtype=tf.float32))[0][0]
+
+        if desired_class == "opposite":
+            desired_class = 1.0 - np.round(test_pred)
+        else:
+            desired_class = np.round(test_pred)
+        query_instance_copy = query_instance.copy()
+
+        #preparing query instance for conversion to pandas dataframe
+        for q in query_instance_copy:
+            query_instance_copy[q] = [query_instance_copy[q]]
+        query_instance_df = pd.DataFrame.from_dict(query_instance_copy)
+
+        start_time = timeit.default_timer()
+
+        #Making the one-hot-encoded version of query instance match the one-hot encoded version of the dataset
+        query_instance_df_dummies = pd.get_dummies(query_instance_df)
+        for col in pd.get_dummies(self.data_interface.data_df[self.data_interface.feature_names]).columns:
+            if col not in query_instance_df_dummies.columns:
+                query_instance_df_dummies[col] = 0
+
+        # Finding counterfactuals from the KD Tree
+        indices = self.KD_tree[desired_class].query(query_instance_df_dummies, self.total_CFs)[1][0].tolist()
+        final_cfs = self.dataset_with_predictions[desired_class][self.data_interface.feature_names].iloc[indices].copy()
+
+        #finding the predicted outcome for each cf
+        self.cfs_preds = []
+        for i in range(len(indices)):
+            cfs = final_cfs.iloc[i].copy()
```

Comment:
again, the code leads to conversion from df->dict->df. not needed, try to directly work with the df. the prepare_query_instance assumed dict input from the user, that's why its input is a dict.